dbcon/execplan/treenode.h

```diff
@@ -1031,41 +1036,83 @@ inline IDB_Decimal TreeNode::getDecimalVal()
         case CalpontSystemCatalog::INT:
         case CalpontSystemCatalog::SMALLINT:
         case CalpontSystemCatalog::TINYINT:
-            fResult.decimalVal.value = (int64_t)(fResult.intVal * pow((double)10.0, fResultType.scale));
-            fResult.decimalVal.scale = fResultType.scale;
-            fResult.decimalVal.precision = fResultType.precision;
+        {
+            idbassert(fResultType.scale == 0);
+
+            fResult.decimalVal = IDB_Decimal(fResult.intVal, 0,
+                                     fResultType.precision);
+
             break;
+        }
 
         case CalpontSystemCatalog::UBIGINT:
+        {
+            idbassert(fResultType.scale == 0);
+
+            if (fResultType.isWideDecimalPrecision())
+                fResult.decimalVal = IDB_Decimal(0, 0, fResultType.precision,
+                                         (int128_t)fResult.uintVal);
+            else
+                fResult.decimalVal = IDB_Decimal((int64_t)fResult.uintVal, 0,
+                                         fResultType.precision);
+
+            break;
+        }
+
         case CalpontSystemCatalog::UMEDINT:
         case CalpontSystemCatalog::UINT:
         case CalpontSystemCatalog::USMALLINT:
         case CalpontSystemCatalog::UTINYINT:
-            fResult.decimalVal.value = (int64_t)(fResult.uintVal * pow((double)10.0, fResultType.scale));
-            fResult.decimalVal.scale = fResultType.scale;
-            fResult.decimalVal.precision = fResultType.precision;
+        {
+            idbassert(fResultType.scale == 0);
+
+            fResult.decimalVal = IDB_Decimal((int64_t)fResult.uintVal, 0,
+                                     fResultType.precision);
+
             break;
+        }
 
         case CalpontSystemCatalog::LONGDOUBLE:
+        {
+            long double dlScaled = fResult.longDoubleVal;
+
+            if (fResultType.scale > 0)
             {
-                long double dlScaled = fResult.longDoubleVal;
-                if (fResultType.scale > 0)
-                {
-                    dlScaled= fResult.longDoubleVal * pow((double)10.0, fResultType.scale);
-                }
-                if (dlScaled > (double)MAX_BIGINT)
-                {
-                    throw logging::InvalidConversionExcept("TreeNode::getDecimalVal: decimal overflow.");
-                }
-                fResult.decimalVal.value = (int64_t)roundl((fResult.longDoubleVal * pow((double)10.0, fResultType.scale)));
-                fResult.decimalVal.scale = fResultType.scale;
-                fResult.decimalVal.precision = fResultType.precision;
+                dlScaled = fResult.longDoubleVal * pow((double)10.0, fResultType.scale);
             }
+
+            if ((dlScaled > (long double)INT64_MAX) || (dlScaled < (long double)(INT64_MIN)))
+            {
+                int128_t temp = datatypes::Decimal::getInt128FromFloat128(
+                                    roundq(__float128(dlScaled)));
```

Comment:
We surely will remove quadmath from the code b/c it brings additional dependency so we shouldn't use it.

Reply:
I have removed the `roundq` call from this block. The narrow decimal version uses a `lroundl` call however.